plugin/src/main/kotlin/me/tylerbwong/gradle/metalava/extension/MetalavaExtension.kt

```diff
@@ -98,4 +98,36 @@ open class MetalavaExtension {
      * The default is "current".
      */
     var releaseType = "current"
+
+    /**
+     * The directories to search for source files. An exception will be thrown is the named
+     * directories are not direct children of the project root. The default is "src".
+     */
+    var sourcePaths = mutableSetOf("src")
+
+    /**
+     * The sub-directories of those specified by [sourcePaths] that contain any of the given strings
+     * that are not to be searched for source files. The default is "test".
+     *
+     * The behavior of this property can be configured using [ignoreSourcePathsExactMatch].
+     *
+     * @see ignoreSourcePathsExactMatch
+     */
+    var ignoreSourcePaths = mutableSetOf("test")
+
+    /**
+     * Flag to be used in conjunction with [ignoreSourcePaths] to toggle the matching behaviour. The
+     * default behavior (i.e. `false`) is that directories specified by [ignoreSourcePaths] are
+     * successfully matched if the elements of [ignoreSourcePaths] form (using a case-insensitive
+     * match) part of the directory name (i.e. `String.contains(ignoreCase = true)`). Enabling this
+     * flag will change that behavior to case-sensitive exact-matching (i.e. equality) only.
+     *
+     * @see ignoreSourcePaths
+     */
+    var ignoreSourcePathsExactMatch = false
```

Comment:
In order to maintain backward compat, but permit the stricter matching on the dir names (ie using equals(), rather than contains()), this flag will toggle that behaviour in conjunction with the `ignoreSourcePaths` property.

Reply:
Instead of allowing loose matches with `contains`, I think personally I would prefer consumers specify _exactly_ which directories they want ignored. Although it is documented well here, it feels like the expected behavior would the a more strict `equals`. I would be ok with an `equals(..., ignoreCase = true)`, but I think that's as lenient as we should go. So perhaps this flag is not needed.

Reply:
The current behaviour, on `master` is that the plugin conducts a loose match (using `contains()`), and if that was to change to be stricter (using any form of `equals()`) then all existing users of this plugin would be broken.

I completely agree that exact matching is the way to go, but in order to support that behaviour the old matching needs to keep working as the default, with the new, stricter, behaviour an opt-in.

It wouldn't be entirely unreasonable to deprecate the old behaviour, though, and log a warning that this default behaviour is to be removed at some point (ie giving users some time to integrate the strict behaviour). Because the behaviour is default, though, one can't just annotate the property with `@Deprecated`, as direct access to that won't typically be used, and so, as mentioned, a warning would need to be logged at runtime.

Thus, a proposal for doing this would be:

1. Introduce both new properties (`ignoreSourcePaths` and `ignoreSourcePathsExactMatch`).
2. Log a runtime warning that the default behaviour of loose matching will be removed, and that `ignoreSourcePaths` and `ignoreSourcePathsExactMatch = true` should be used.
3. Release a new plugin version with a MINOR bump.
4. Remove the loose matching behaviour, the aforementioned log message, and explicitly `@Deprecate` the `ignoreSourcePathsExactMatch` property (as everyone should then be using it anyway).
5. Release a new plugin version with a MAJOR bump.
6. Remove the `ignoreSourcePathsExactMatch` property
7. Release a new plugin version with a MAJOR bump.

Presumably, all these updates can be done sequentially, at the same time, as the option is on the User to choose to integrate the plugin when there's a MAJOR version change?

Of course, that's if you wish to maintain compatibility going forward. An alternative would be to break the functionality by removing the loose matching behaviour, and publish.

Reply:
I'm not too concerned with maintaining compatibility with this moving forward. The `contains` catch-all was mainly in place to cover all the multiplatform possibilities for test src directories (`commonTest`, `iosTest`, `jsTest`, etc.), so maybe we can just hardcode those as defaults and leave it be. It seems minor enough to where as long as it's documented in the release notes we should be good.